### ISSUE TITLE
Accessibility for limited vision users

### ISSUE THREAD

**Issue by INFU-AV** [tommitytom/RetroPlug#17]
Thanks to RetroPlug LSDj could be easily designed to be accessible if reading out names of parameters and cursor position would be available to visually impared, and/or through some other solution compatible with open source screen reader, NVDA!

**Comment by juanreina921** [comment]
Hey, my name is Juan, and I also think there could be a way for screen reader accessibility, especially for those who have less vision and those like myself, who have no vision at all. I am also using the open sorce screen reader NVDA, and am a hardcore chiptune fanatic. Maybe like what @INFU-AV said, we could have the screen reader speak cursor position, speak effects an parameters,or even speak something like Alert., Lsdj has started. Upon starting of lsdj. @tommitytom I am not a coder by any means, but I would suggest you to grab the screen reader from here.
https://www.nvaccess.org/download/
This allows you to get better familiar with the screen reader, so you can get a feal for how it operates. And lastly, there is a surtain file that is distributed with some games, and other software like retro arch. That file is known as NVDA controler client32 or NVDA controler client64.dll and is located in program files or program files x 86\NVDA. and if I am correct, it allows the UI to be more visible to NVDA.
Only two downsides are these files must be distributed with the program for eas of use, and that you'll have to change the code for the UI so NVDA could take notice.
I hope this helps, and we hope to here your thoughts and ideas on this issue.

**Comment by tommitytom** [comment]
This is incredibly interesting!  It's something I would definitely like to implement further down the line, but will be quite a difficult task since it will require parsing the screen output from LSDj (RetroPlug currently has no idea what screen is being viewed, or where the cursor is on the screen etc).  I'm currently working on implementing scripting functionality that should make tasks like this a lot easier, so I will revisit this when that has progressed further :)

**Comment by cyberic99** [comment]
> RetroPlug currently has no idea what screen is being viewed

I played with the bgb debugger running LSDJ, and I noticed that the current patterns, cursors etc, were located at a fixed offset and could be easily grabbed from memory.

This could be easier than reading the screen (which could be quite hard with custom fonts and colors)

**Comment by tommitytom** [comment]
@cyberic99 Yes I'll be adding the ability to access different sections of GB memory from within lua so that will most likely be the approach I take.  Thanks for the heads up!

**Comment by tommitytom** [comment]
HI @juanreina921, is this something you're still interested in? TTS has come a long way since this issue was opened (especially AI based ones), and it's a lot easier for me to embed. However, you'll have to excuse my ignorance, but I'm unaware of what kinds of TTS are most useful for limited vision users. Any guidance would be welcome!

**Comment by juanreina921** [comment]
Well hello! Yes I am still very much interested in this, but if I may, I think you're thinking too big into this, at least with regards to the tts system. 
The primary focus wasn't to get some or any AI tts systems to read/give information to the user that being the blind and visually impaired, especially as we already have more than capable tts systems that are on our screen readers that we can utilize and control the force of information with, but what we don't have is the structured information itself. 

Let's just take the latest standalone Windos version of RetroPlug as that's the most readable version at the moment. 

Now when you open the latest win version of RetroPlug that being the version from 2022 using NVDA you should get the window title and version, press alt and you should be in the file menu. 
From the file menu press down and you should see 2 options: Preferences; and quit. 
In the preferences if you use the Tab key you should see 3 options or there abouts: Okay; cancel; and aply. Now this is 3 pluss years old and so may have not been fully worked out on the keyboard but where are the actual preferences at? Well by total chance I've found that if you press either the rite or left arrow key a combo box may just appear from thin air, but don't press Tab because that option will just go away and also that combo box will be the only thing you can work with as the other options if they're any they will not show up in the UI, as I haven't found how to make them appear. 

This might sound really disrespectful but I assure you I don't mean it like that, but please don't just slap AI over it because this isn't correct structure, never mind getting LSDJ to be readable for now. I saw that you've done a rewrite, but please don't tell me that you've changed the UI toolkit to something like IMGUI or something, because then we'll have a big mess on our hands. I've put in word with the Author of IMGUI and also got a blind programmer on the matter to help if any but as of now IMGUI'S accessibility work has yet to move forward. 

I'm not totally against the uce of AI you know, but all I'm saying is you need to learn the accessibility structure at least. I'm not trying to get you to use a screen reader full time na none of that, but before we move forward with any kind of hacky solutes let's at least see if we can fix the standalone versions at least, we can worry about LSDJ later. 

I'm not a programmer but I am someone who is finally making music in a tracker now, and it was done via the work of blind and visually impaired people and the OpenMPT team, but we did it with structure and foundation and not just AIing it up and letting it fall aside. 


So yes of course I am interested in this! But are you interested in learning a little bit of structure? I can and will find all the documents I can and if you wish I will be fully available over voice if you need something.
